apps/emqx_durable_storage/src/emqx_ds_replication_layer_shard.erl

```diff
@@ -84,15 +83,14 @@ server_name(DB, Shard, Site) ->
 %%
 
 -spec servers(emqx_ds:db(), emqx_ds_replication_layer:shard_id(), Order) -> [server(), ...] when
-    Order :: leader_preferred | undefined.
-servers(DB, Shard, _Order = leader_preferred) ->
+    Order :: leader_preferred | local_preferred | undefined.
+servers(DB, Shard, leader_preferred) ->
```

Comment:
Just to be sure we're aware of that: this works only for shards that have a replica on the local node, because `ra_leaderboard` does not track leaders otherwise (i.e. when all of the servers are supposed to live on remote nodes).

Reply:
Does it track members at least?

Reply:
Leaderboard tracks members + leaders for all shards that have local replicas. Otherwise `get_shard_servers/2` is called that simply lists currently allocated shard servers, which obviously lacks leadership information.

Reply:
I think we need a more generic mechanism for shooting requests to the leader. It'll help to split storing batches into prepare/commit stages as well. I'll leave it as is for now, if you don't object.